src/app-layer-smb.c

```diff
@@ -36,11 +39,145 @@ static SuricataFileContext sfc = { &sbcfg };
+
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        return;
+    ipp = IPPairGetIPPairFromHash(&ip_src, &ip_dst);
+    if (ipp == NULL)
+        return;
+
+    SmbMultiIppairKey *key = IPPairGetStorageById(ipp, g_ippair_smb_multi_id);
+
+    if (key) {
+        // already a key
+        if (memcmp(guid, key->guid, GUID_LEN) == 0) {
+            // same guid, update only eof
+            if (eof > key->filesize) {
+                key->filesize = eof;
+            }
+            if (name_len > 0 && key->filename == NULL) {
+                key->filename = SCMalloc(name_len);
+                if (key->filename == NULL) {
+                    SCFree(key);
+                    goto end;
+                }
+                key->filename_len = name_len;
+                memcpy(key->filename, filename, name_len);
+            }
+            goto end;
+        } else {
+            // different guid : keep old one TODOsmbmulti6 : handle multimulti ?
+            goto end;
+        }
+    }
+    // else
+    key = SCCalloc(1, sizeof(SmbMultiIppairKey));
+    if (key == NULL) {
+        goto end;
+    }
+
+    memcpy(key->guid, guid, GUID_LEN);
+    key->filesize = eof;
+    if (name_len > 0) {
+        key->filename = SCMalloc(name_len);
+        if (key->filename == NULL) {
+            SCFree(key);
+            goto end;
+        }
+        key->filename_len = name_len;
+        memcpy(key->filename, filename, name_len);
+    }
+    IPPairSetStorageById(ipp, g_ippair_smb_multi_id, key);
+end:
+    IPPairUnlock(ipp);
+}
+
+#define SMB_URL_PREFIX_LEN 6
+
+HttpRangeContainerBlock *SmbMultiStartFileChunk(const Flow *f, const uint8_t *guid, uint16_t flags,
+        FileContainer *fc, const StreamingBufferConfig *sbcfg, bool *added, uint64_t offset,
+        uint32_t rlen, const uint8_t *data, uint32_t data_len)
+{
+    IPPair *ipp;
+    Address ip_src, ip_dst;
+    HttpRangeContainerBlock *r = NULL;
+
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        return NULL;
+    ipp = IPPairLookupIPPairFromHash(&ip_src, &ip_dst);
+    if (ipp == NULL)
+        return NULL;
+
+    SmbMultiIppairKey *key = IPPairGetStorageById(ipp, g_ippair_smb_multi_id);
+    if (key == NULL) {
+        key = SCCalloc(1, sizeof(SmbMultiIppairKey));
+        if (key == NULL) {
+            goto end;
+        }
+        memcpy(key->guid, guid, GUID_LEN);
+        key->filesize = offset + rlen;
+        IPPairSetStorageById(ipp, g_ippair_smb_multi_id, key);
+    } else if (memcmp(guid, key->guid, GUID_LEN) != 0) {
+        goto end;
+    }
+    FileContentRange fcr;
+    if (offset > INT64_MAX || offset >= key->filesize || key->filesize > INT64_MAX) {
+        goto end;
+    }
+    fcr.start = offset;
+    fcr.size = key->filesize;
+    if (offset + rlen > key->filesize) {
+        fcr.end = key->filesize;
+    } else {
+        fcr.end = offset + rlen;
+    }
+    // TODOsmbmulmti4 should we only rely on ippair and have nothing global ?
```

Comment:
That is the big question for me...

Reply:
Not sure. What do the SMB protocol docs have to say about this?

Reply:
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb/44c3cf8d-0931-4923-8fdc-738537ba70ba

>   The FileId MUST be unique for a file on a given object store.

So, look like it does not depend on the client, only the server...

But SMB2 is less clear 
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-smb2/f1d9b40d-e335-45fc-9d0b-199a31ede4c3

So, it is maybe best to not use ippair, and use a global hash table as for HTTP ranges, with the key having the server address and the file unique id...

Reply:
Might be good to look at other tools to see how they solve this. Zeek and Wireshark come to mind. Samba might also be interesting to look at.